Source/Core/Core/IOS/Network/IP/Top.cpp

```diff
@@ -162,70 +167,223 @@ static s32 MapWiiSockOptNameToNative(u32 optname)
   return optname;
 }
 
-// u32 values are in little endian (i.e. 0x0100007f means 127.0.0.1)
+struct InterfaceRouting
+{
+  u32 index;
+  in_addr destination;
+  in_addr netmask;
+  in_addr gateway;
+};
+
 struct DefaultInterface
 {
-  u32 inet;       // IPv4 address
-  u32 netmask;    // IPv4 subnet mask
-  u32 broadcast;  // IPv4 broadcast address
+  in_addr inet;                                 // IPv4 address
+  in_addr netmask;                              // IPv4 subnet mask
+  in_addr broadcast;                            // IPv4 broadcast address
+  std::vector<InterfaceRouting> routing_table;  // IPv4 routing table
 };
 
-static std::optional<DefaultInterface> GetSystemDefaultInterface()
+static std::vector<InterfaceRouting> GetSystemInterfaceRouting()
 {
+  std::vector<InterfaceRouting> routing_table;
+
 #ifdef _WIN32
-  std::unique_ptr<MIB_IPFORWARDTABLE> forward_table;
   DWORD forward_table_size = 0;
+  std::unique_ptr<MIB_IPFORWARDTABLE> forward_table;
   if (GetIpForwardTable(nullptr, &forward_table_size, FALSE) == ERROR_INSUFFICIENT_BUFFER)
   {
     forward_table =
         std::unique_ptr<MIB_IPFORWARDTABLE>((PMIB_IPFORWARDTABLE) operator new(forward_table_size));
   }
 
-  std::unique_ptr<MIB_IPADDRTABLE> ip_table;
-  DWORD ip_table_size = 0;
-  if (GetIpAddrTable(nullptr, &ip_table_size, FALSE) == ERROR_INSUFFICIENT_BUFFER)
-  {
-    ip_table = std::unique_ptr<MIB_IPADDRTABLE>((PMIB_IPADDRTABLE) operator new(ip_table_size));
-  }
-
-  // find the interface IP used for the default route and use that
-  NET_IFINDEX ifIndex = NET_IFINDEX_UNSPECIFIED;
   DWORD result = GetIpForwardTable(forward_table.get(), &forward_table_size, FALSE);
   // can return ERROR_MORE_DATA on XP even after the first call
   while (result == NO_ERROR || result == ERROR_MORE_DATA)
   {
-    for (DWORD i = 0; i < forward_table->dwNumEntries; ++i)
+    const std::span<MIB_IPFORWARDROW> table(forward_table->table, forward_table->dwNumEntries);
+    for (const auto& entry : table)
     {
-      if (forward_table->table[i].dwForwardDest == 0)
+      routing_table.emplace_back(entry.dwForwardIfIndex,
+                                 std::bit_cast<in_addr>(entry.dwForwardDest),
+                                 std::bit_cast<in_addr>(entry.dwForwardMask),
+                                 std::bit_cast<in_addr>(entry.dwForwardNextHop));
+    }
+
+    if (result == NO_ERROR)
+      break;
+
+    result = GetIpForwardTable(forward_table.get(), &forward_table_size, FALSE);
+  }
+
+#elif defined(__linux__)
+  constexpr int BUFF_SIZE = 8192;
+  unsigned int msg_seq = 0;
+  const int sock = socket(PF_NETLINK, SOCK_DGRAM, NETLINK_ROUTE);
+  if (sock < 0)
+  {
+    ERROR_LOG_FMT(IOS_NET, "Failed to open netlink socket, error {}", Common::StrNetworkError());
+    return {};
+  }
+
```

Comment:
A read timeout should be added to the socket as it can be blocking indefinitely in some circumstances (e.g. packet loss). I managed to produce a similar result by setting `nlmsg_flags = 0`.

```C++
  const timeval tv{.tv_sec = 2, .tv_usec = 0};
  if (setsockopt(sock, SOL_SOCKET, SO_RCVTIMEO, static_cast<const void*>(&tv), sizeof(tv)) < 0)
  {
    ERROR_LOG_FMT(IOS_NET, "Failed to set netlink socket recv timeout: {}", Common::StrNetworkError());
    return {};
  }
```